# HW3: NER с трансформерными моделями

**Задача:** Дообучение трансформерной модели для NER-задачи на русском языке

**Датасет:** factRuEval-2016 (gusevski/factrueval2016)

**Базовая модель:** cointegrated/rubert-tiny2

---

## План работы

1. **Базовое обучение NER** — загрузка данных, препроцессинг, baseline
2. **MLM предобучение** — адаптация модели к домену через Masked Language Modeling
3. **Улучшенное маскирование** — WholeWordMask и Entity masking
4. **Синтетическая разметка** — генерация pseudo-labels через DeepPavlov NER
5. **Сравнение подходов** — финальные выводы

---

## 0. Настройка окружения и импорты

In [1]:
# Установка зависимостей
# --- Google Colab / Kaggle (активная установка) ---
# Версии зафиксированы для воспроизводимости: на этой версии transformers
# работает eval_strategy, корректно принимается tokenizer= в Trainer,
# и DataCollatorForWholeWordMask ведёт себя как ожидается.
!pip install -q "transformers==4.44.2" "datasets==2.21.0" seqeval corus accelerate huggingface_hub

# --- Локальный uv-проект (выполнить в терминале, не в ноутбуке) ---
# uv add "transformers==4.44.2" "datasets==2.21.0" seqeval corus accelerate huggingface_hub


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 2.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 3.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 92.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.3/527.3 kB 44.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 83.7/83.7 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 47.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.6/177.6 kB 19.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 123.3 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2025.3.0 requires fsspec==2025.3.0, but you have fsspec 2024.6.1 which is incompatible.


In [11]:
# Фиксируем random_state для воспроизводимости
RANDOM_STATE = 42

import warnings
warnings.filterwarnings('ignore')

import os
import numpy as np
import pandas as pd
from tqdm.auto import tqdm
from collections import Counter

# Фиксируем seeds
np.random.seed(RANDOM_STATE)

import torch
torch.manual_seed(RANDOM_STATE)

from torch.utils.data import DataLoader

def get_device():
    """
    Автоматическое определение устройства.
    Приоритет: CUDA -> MPS (Apple Silicon) -> CPU
    """
    if torch.cuda.is_available():
        device = torch.device('cuda')
        torch.cuda.manual_seed_all(RANDOM_STATE)
        print(f'GPU: {torch.cuda.get_device_name(0)}')
    elif hasattr(torch.backends, 'mps') and torch.backends.mps.is_available():
        device = torch.device('mps')
        print('Using Apple Silicon MPS')
    else:
        device = torch.device('cpu')
        print('Using CPU')
    return device

DEVICE = get_device()

# Параметры обучения, адаптированные под устройство
if DEVICE.type == 'cuda':
    BATCH_SIZE = 16
    EVAL_BATCH_SIZE = 32
    FP16 = True  # mixed precision для GPU
elif DEVICE.type == 'mps':
    BATCH_SIZE = 8
    EVAL_BATCH_SIZE = 16
    FP16 = False  # MPS не поддерживает fp16 полноценно
else:  # CPU
    BATCH_SIZE = 8
    EVAL_BATCH_SIZE = 8
    FP16 = False

print(f'Device: {DEVICE}')
print(f'Batch size: {BATCH_SIZE}, Eval batch size: {EVAL_BATCH_SIZE}, FP16: {FP16}')

GPU: Tesla T4
Device: cuda
Batch size: 16, Eval batch size: 32, FP16: True


In [12]:
from datasets import load_dataset, DatasetDict, Dataset
from transformers import (
    AutoTokenizer,
    AutoModelForTokenClassification,
    AutoModelForMaskedLM,
    DataCollatorForTokenClassification,
    DataCollatorForLanguageModeling,
    DataCollatorForWholeWordMask,
    TrainingArguments,
    Trainer,
    set_seed
)
from seqeval.metrics import (
    classification_report as seqeval_report,
    f1_score as seqeval_f1,
    precision_score as seqeval_precision,
    recall_score as seqeval_recall
)

set_seed(RANDOM_STATE)

## 1. Загрузка и анализ датасета

**factRuEval-2016** — эталонный корпус для оценки NER на русском языке.

Сущности:
- **PER** (Person) — имена людей
- **ORG** (Organization) — названия организаций  
- **LOC** (Location) — географические объекты

In [13]:
# Загрузка датасета
# Датасет имеет нестандартную структуру: данные упакованы в поле 'data' как список словарей
raw_dataset = load_dataset('gusevski/factrueval2016')
print('Исходная структура:')
print(raw_dataset)

# Распаковываем данные из поля 'data' в нормальный формат
def unpack_dataset(raw_split):
    """Распаковка данных из поля 'data' в отдельные примеры."""
    # data[0] содержит список всех примеров
    examples = raw_split['data'][0]
    return Dataset.from_list(examples)

dataset = DatasetDict({
    'train': unpack_dataset(raw_dataset['train']),
    'validation': unpack_dataset(raw_dataset['validation']),
    'test': unpack_dataset(raw_dataset['test'])
})

print('\nРаспакованная структура:')
print(dataset)

Repo card metadata block was not found. Setting CardData to empty.


Исходная структура:
DatasetDict({
    train: Dataset({
        features: ['data'],
        num_rows: 1
    })
    validation: Dataset({
        features: ['data'],
        num_rows: 1
    })
    test: Dataset({
        features: ['data'],
        num_rows: 1
    })
})

Распакованная структура:
DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'length', 'ner_tags_str', 'ner_tags'],
        num_rows: 7746
    })
    validation: Dataset({
        features: ['id', 'tokens', 'length', 'ner_tags_str', 'ner_tags'],
        num_rows: 2582
    })
    test: Dataset({
        features: ['id', 'tokens', 'length', 'ner_tags_str', 'ner_tags'],
        num_rows: 2582
    })
})


In [14]:
# Анализ структуры данных
print('Пример из train:')
example = dataset['train'][0]
print(f"ID: {example['id']}")
print(f"Tokens ({len(example['tokens'])}): {example['tokens'][:10]}...")
print(f"NER tags (str): {example['ner_tags_str'][:10]}...")
print(f"NER tags (int): {example['ner_tags'][:10]}...")

Пример из train:
ID: 0
Tokens (47): ['"', 'Если', 'Миронов', 'занял', 'столь', 'оппозиционную', 'позицию', ',', 'то', 'мне']...
NER tags (str): ['O', 'O', 'B-PER', 'O', 'O', 'O', 'O', 'O', 'O', 'O']...
NER tags (int): [0, 0, 1, 0, 0, 0, 0, 0, 0, 0]...


In [15]:
# Получаем список меток из данных
# Собираем уникальные метки из ner_tags_str
all_tags = set()
for example in dataset['train']:
    all_tags.update(example['ner_tags_str'])

# Сортируем метки: O первым, затем B-*, I-* в алфавитном порядке
label_list = sorted(all_tags, key=lambda x: (x != 'O', x))
print(f'Метки NER ({len(label_list)}): {label_list}')

# Создаём маппинги
label2id = {label: i for i, label in enumerate(label_list)}
id2label = {i: label for i, label in enumerate(label_list)}

print(f'\nlabel2id: {label2id}')

Метки NER (7): ['O', 'B-LOC', 'B-ORG', 'B-PER', 'I-LOC', 'I-ORG', 'I-PER']

label2id: {'O': 0, 'B-LOC': 1, 'B-ORG': 2, 'B-PER': 3, 'I-LOC': 4, 'I-ORG': 5, 'I-PER': 6}


### BIO-схема разметки

| Метка | Значение | Описание |
|-------|----------|----------|
| **B-PER** | **B**eginning of Person | Начало сущности "персона" |
| **I-PER** | **I**nside Person | Продолжение сущности "персона" |
| **B-ORG** | **B**eginning of Organization | Начало сущности "организация" |
| **I-ORG** | **I**nside Organization | Продолжение сущности "организация" |
| **B-LOC** | **B**eginning of Location | Начало сущности "локация" |
| **I-LOC** | **I**nside Location | Продолжение сущности "локация" |
| **O** | Outside | Не является сущностью |

In [16]:
# Примеры BIO-разметки из датасета
print('ПРИМЕРЫ BIO-РАЗМЕТКИ')
print('='*70)

examples_shown = 0
for example in dataset['train']:
    tokens = example['tokens']
    tags = example['ner_tags_str']

    # Ищем многословные сущности (B-* + I-*)
    i = 0
    while i < len(tags):
        if tags[i].startswith('B-'):
            ent_type = tags[i][2:]  # PER, ORG, LOC
            start_i = i
            i += 1
            # Собираем все I-* токены
            while i < len(tags) and tags[i] == f'I-{ent_type}':
                i += 1
            end_i = i

            # Показываем только многословные сущности (минимум 2 токена)
            if end_i - start_i >= 2 and examples_shown < 8:
                # Контекст: 2 слова до и после
                ctx_start = max(0, start_i - 2)
                ctx_end = min(len(tokens), end_i + 2)

                print(f'\nСущность: "{" ".join(tokens[start_i:end_i])}" ({ent_type})')
                print(f'Токены:   {tokens[start_i:end_i]}')
                print(f'Метки:    {tags[start_i:end_i]}')
                print(f'Контекст: ...{" ".join(tokens[ctx_start:ctx_end])}...')
                examples_shown += 1
        else:
            i += 1

    if examples_shown >= 8:
        break

print('\n' + '='*70)
print('Каждая сущность начинается с B-*, продолжается I-* того же типа')

ПРИМЕРЫ BIO-РАЗМЕТКИ

Сущность: "Единой Россией" (ORG)
Токены:   ['Единой', 'Россией']
Метки:    ['B-ORG', 'I-ORG']
Контекст: ...сегодня " Единой Россией '' и...

Сущность: "Ханты-Мансийском автономном округе" (LOC)
Токены:   ['Ханты-Мансийском', 'автономном', 'округе']
Метки:    ['B-LOC', 'I-LOC', 'I-LOC']
Контекст: ...В Ханты-Мансийском автономном округе с должности...

Сущность: "Николай Гудожников" (PER)
Токены:   ['Николай', 'Гудожников']
Метки:    ['B-PER', 'I-PER']
Контекст: ...начальник УВД Николай Гудожников ....

Сущность: "Российского научного центра" (ORG)
Токены:   ['Российского', 'научного', 'центра']
Метки:    ['B-ORG', 'I-ORG', 'I-ORG']
Контекст: ...является президентом Российского научного центра " Курчатовский...

Сущность: "Красненькая речка" (LOC)
Токены:   ['Красненькая', 'речка']
Метки:    ['B-LOC', 'I-LOC']
Контекст: ...округе " Красненькая речка '' ....

Сущность: "Карл Айкенберри" (PER)
Токены:   ['Карл', 'Айкенберри']
Метки:    ['B-PER', 'I-PER']
Контекст: ...

In [17]:
# Статистика по сущностям: считаем и токены, и сущности

def count_statistics(dataset_split):
    """
    Подсчёт статистики по сущностям в датасете.

    Возвращает:
    - entity_counts: количество сущностей (считаем только B-* теги)
    - token_counts: количество токенов с разметкой (B-* и I-*)
    """
    entity_counts = Counter()  # количество сущностей
    token_counts = Counter()   # количество токенов

    for example in dataset_split:
        for tag in example['ner_tags_str']:
            if tag.startswith('B-'):
                # Начало новой сущности
                entity_type = tag[2:]  # B-PER -> PER
                entity_counts[entity_type] += 1
                token_counts[entity_type] += 1
            elif tag.startswith('I-'):
                # Продолжение сущности (только токен, не новая сущность)
                entity_type = tag[2:]  # I-PER -> PER
                token_counts[entity_type] += 1

    return entity_counts, token_counts

print('СТАТИСТИКА ПО ДАТАСЕТУ')
print('='*70)

for split in dataset:
    entity_counts, token_counts = count_statistics(dataset[split])
    total_entities = sum(entity_counts.values())
    total_tokens = sum(token_counts.values())

    print(f'\n{split.upper()} ({len(dataset[split])} примеров)')
    print('-'*50)
    print(f'{"Тип":<6} {"Сущностей":>12} {"Токенов":>12} {"Ср. длина":>12}')
    print('-'*50)

    for ent_type in ['PER', 'ORG', 'LOC']:
        ent_count = entity_counts[ent_type]
        tok_count = token_counts[ent_type]
        avg_len = tok_count / ent_count if ent_count > 0 else 0
        print(f'{ent_type:<6} {ent_count:>12} {tok_count:>12} {avg_len:>12.2f}')

    print('-'*50)
    total_avg = total_tokens / total_entities if total_entities > 0 else 0
    print(f'{"ВСЕГО":<6} {total_entities:>12} {total_tokens:>12} {total_avg:>12.2f}')

СТАТИСТИКА ПО ДАТАСЕТУ

TRAIN (7746 примеров)
--------------------------------------------------
Тип       Сущностей      Токенов    Ср. длина
--------------------------------------------------
PER            6359        10284         1.62
ORG            5084         8311         1.63
LOC            4335         5307         1.22
--------------------------------------------------
ВСЕГО         15778        23902         1.51

VALIDATION (2582 примеров)
--------------------------------------------------
Тип       Сущностей      Токенов    Ср. длина
--------------------------------------------------
PER            2110         3431         1.63
ORG            1629         2815         1.73
LOC            1394         1701         1.22
--------------------------------------------------
ВСЕГО          5133         7947         1.55

TEST (2582 примеров)
--------------------------------------------------
Тип       Сущностей      Токенов    Ср. длина
-----------------------------------------

---

## 2. Подготовка данных для обучения

### Выбор базовой модели

**cointegrated/rubert-tiny2** — компактный BERT для русского языка:
- 12 слоёв, 312 hidden size
- ~29M параметров (в 6 раз меньше BERT-base)
- Быстрое обучение и инференс
- Хорошее качество для своего размера

In [18]:
# Используем компактную модель для быстрого обучения
MODEL_NAME = 'cointegrated/rubert-tiny2'

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
print(f'Tokenizer: {tokenizer.__class__.__name__}')
print(f'Vocab size: {tokenizer.vocab_size}')

Tokenizer: BertTokenizerFast
Vocab size: 83828


### Токенизация с выравниванием меток

BERT использует subword токенизацию (WordPiece). Одно слово может разбиться на несколько токенов:

```
"Путин" -> ["Пу", "##тин"]
```

**Стратегия выравнивания:**
- Первый субтокен слова получает метку слова (B-PER, I-ORG и т.д.)
- Остальные субтокены получают `-100` (игнорируются при расчёте loss)
- Специальные токены [CLS], [SEP] также получают `-100`

In [19]:
def tokenize_and_align_labels(examples):
    """
    Токенизация с выравниванием NER-меток.

    Стратегия:
    - Первый субтокен слова получает метку слова
    - Остальные субтокены получают -100 (игнорируются в loss)
    - Специальные токены [CLS], [SEP] получают -100
    """
    tokenized = tokenizer(
        examples['tokens'],
        truncation=True,
        is_split_into_words=True,  # входные данные уже токенизированы по словам
        max_length=512
    )

    labels = []
    for i, ner_tags_str in enumerate(examples['ner_tags_str']):
        word_ids = tokenized.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids_aligned = []

        for word_idx in word_ids:
            if word_idx is None:
                # Специальные токены [CLS], [SEP], [PAD]
                label_ids_aligned.append(-100)
            elif word_idx != previous_word_idx:
                # Первый субтокен слова — конвертируем строковую метку в ID
                tag_str = ner_tags_str[word_idx]
                label_ids_aligned.append(label2id[tag_str])
            else:
                # Продолжение слова (##token) — игнорируем
                label_ids_aligned.append(-100)

            previous_word_idx = word_idx

        labels.append(label_ids_aligned)

    tokenized['labels'] = labels
    return tokenized

In [20]:
# Применяем токенизацию к датасету
tokenized_dataset = dataset.map(
    tokenize_and_align_labels,
    batched=True,
    remove_columns=dataset['train'].column_names
)

print(tokenized_dataset)

Map:   0%|          | 0/7746 [00:00<?, ? examples/s]

Map:   0%|          | 0/2582 [00:00<?, ? examples/s]

Map:   0%|          | 0/2582 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'labels'],
        num_rows: 7746
    })
    validation: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'labels'],
        num_rows: 2582
    })
    test: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'labels'],
        num_rows: 2582
    })
})


In [21]:
# Проверяем выравнивание на примере
example_idx = 0
original = dataset['train'][example_idx]
tokenized = tokenized_dataset['train'][example_idx]

print('Исходные токены и метки:')
for token, tag in zip(original['tokens'], original['ner_tags_str']):
    print(f'  {token:20} -> {tag}')

print('\nПосле токенизации (субтокены):')
tokens = tokenizer.convert_ids_to_tokens(tokenized['input_ids'])
for token, label in zip(tokens[:30], tokenized['labels'][:30]):
    label_str = id2label[label] if label != -100 else '[IGNORE]'
    print(f'  {token:20} -> {label_str}')

Исходные токены и метки:
  "                    -> O
  Если                 -> O
  Миронов              -> B-PER
  занял                -> O
  столь                -> O
  оппозиционную        -> O
  позицию              -> O
  ,                    -> O
  то                   -> O
  мне                  -> O
  представляется       -> O
  ,                    -> O
  что                  -> O
  для                  -> O
  него                 -> O
  было                 -> O
  бы                   -> O
  порядочным           -> O
  и                    -> O
  правильным           -> O
  уйти                 -> O
  в                    -> O
  отставку             -> O
  с                    -> O
  занимаемого          -> O
  им                   -> O
  поста                -> O
  ,                    -> O
  поста                -> O
  ,                    -> O
  который              -> O
  предоставлен         -> O
  ему                  -> O
  сегодня              -> O
  "                

---

## 3. Baseline: NER без дообучения и с дообучением

### 3.1 Метрики оценки

Для NER используем **seqeval** — библиотеку для оценки последовательностей:
- **Precision** — доля правильно предсказанных сущностей среди всех предсказанных
- **Recall** — доля правильно найденных сущностей среди всех истинных
- **F1** — гармоническое среднее precision и recall

In [22]:
def compute_metrics(eval_pred):
    """
    Вычисление NER-метрик с использованием seqeval.
    """
    predictions, labels = eval_pred
    predictions = np.argmax(predictions, axis=2)

    # Убираем игнорируемые токены (-100) и конвертируем в строки
    true_labels = []
    true_predictions = []

    for pred_seq, label_seq in zip(predictions, labels):
        true_label_seq = []
        true_pred_seq = []

        for pred, label in zip(pred_seq, label_seq):
            if label != -100:
                true_label_seq.append(id2label[label])
                true_pred_seq.append(id2label[pred])

        true_labels.append(true_label_seq)
        true_predictions.append(true_pred_seq)

    return {
        'precision': seqeval_precision(true_labels, true_predictions),
        'recall': seqeval_recall(true_labels, true_predictions),
        'f1': seqeval_f1(true_labels, true_predictions)
    }

In [23]:
def evaluate_model(model, dataset_split, description='Evaluating'):
      model.eval()
      all_predictions = []
      all_labels = []

      # Определяем устройство модели
      model_device = next(model.parameters()).device

      data_collator = DataCollatorForTokenClassification(tokenizer)
      dataloader = DataLoader(
          dataset_split,
          batch_size=EVAL_BATCH_SIZE,
          collate_fn=data_collator
      )

      with torch.no_grad():
          for batch in tqdm(dataloader, desc=description):
              # Переносим на устройство МОДЕЛИ
              batch = {k: v.to(model_device) for k, v in batch.items()}
              outputs = model(**batch)

              predictions = outputs.logits.argmax(dim=-1).cpu().numpy()
              labels = batch['labels'].cpu().numpy()

              for pred, label in zip(predictions, labels):
                  valid_indices = label != -100
                  all_predictions.append([id2label[p] for p in pred[valid_indices]])
                  all_labels.append([id2label[l] for l in label[valid_indices]])

      results = {
          'precision': seqeval_precision(all_labels, all_predictions),
          'recall': seqeval_recall(all_labels, all_predictions),
          'f1': seqeval_f1(all_labels, all_predictions)
      }

      print(f"\n{description}")
      print(f"Precision: {results['precision']:.4f}")
      print(f"Recall:    {results['recall']:.4f}")
      print(f"F1:        {results['f1']:.4f}")

      return results


### 3.2 Оценка ДО дообучения (zero-shot)

Загружаем предобученную модель со случайно инициализированной головой классификации и оцениваем качество "из коробки".

In [24]:
# Загружаем модель для NER (со случайной головой классификации)
model_baseline = AutoModelForTokenClassification.from_pretrained(
    MODEL_NAME,
    num_labels=len(label_list),
    id2label=id2label,
    label2id=label2id
).to(DEVICE)

print(f'Модель загружена: {MODEL_NAME}')
print(f'Количество параметров: {sum(p.numel() for p in model_baseline.parameters()):,}')

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Модель загружена: cointegrated/rubert-tiny2
Количество параметров: 29,098,303


In [25]:
# Оценка ДО дообучения
results_before = evaluate_model(
    model_baseline,
    tokenized_dataset['test'],
    'Оценка ДО дообучения (zero-shot)'
)

Оценка ДО дообучения (zero-shot):   0%|          | 0/81 [00:00<?, ?it/s]


Оценка ДО дообучения (zero-shot)
Precision: 0.0156
Recall:    0.0795
F1:        0.0261


### 3.3 Дообучение baseline модели

**Гиперпараметры:**

| Параметр | Значение | Обоснование |
|----------|----------|-------------|
| learning_rate | 2e-5 | Стандарт для fine-tuning BERT |
| batch_size | 8-16 (авто) | Адаптируется под устройство (GPU: 16, CPU/MPS: 8) |
| epochs | 3 | Достаточно для NER на небольшом датасете |
| weight_decay | 0.01 | Регуляризация для предотвращения переобучения |
| warmup_ratio | 0.1 | Плавный старт обучения |
| fp16 | auto | Mixed precision на GPU для ускорения |

In [26]:
# Создаём свежую модель для дообучения
model_finetuned = AutoModelForTokenClassification.from_pretrained(
    MODEL_NAME,
    num_labels=len(label_list),
    id2label=id2label,
    label2id=label2id
).to(DEVICE)

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [27]:
# Настройка обучения
training_args = TrainingArguments(
    output_dir='./results/baseline',
    eval_strategy='epoch',
    save_strategy='epoch',
    learning_rate=2e-5,
    per_device_train_batch_size=BATCH_SIZE,
    per_device_eval_batch_size=EVAL_BATCH_SIZE,
    num_train_epochs=3,
    weight_decay=0.01,
    warmup_ratio=0.1,
    logging_steps=50,
    load_best_model_at_end=True,
    metric_for_best_model='f1',
    fp16=FP16,
    seed=RANDOM_STATE,
    report_to='none',
    use_cpu=(DEVICE.type == 'cpu'),
)

data_collator = DataCollatorForTokenClassification(tokenizer)

In [28]:
trainer = Trainer(
    model=model_finetuned,
    args=training_args,
    train_dataset=tokenized_dataset['train'],
    eval_dataset=tokenized_dataset['test'],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

# Обучение
trainer.train()

Epoch,Training Loss,Validation Loss,Precision,Recall,F1
1,0.155900,0.121710,0.749959,0.833212,0.789396
2,0.093300,0.081633,0.815199,0.887714,0.849913
3,0.079100,0.073792,0.833249,0.896281,0.863616


TrainOutput(global_step=1455, training_loss=0.23004987938297572, metrics={'train_runtime': 53.9443, 'train_samples_per_second': 430.777, 'train_steps_per_second': 26.972, 'total_flos': 18893007289728.0, 'train_loss': 0.23004987938297572, 'epoch': 3.0})

In [29]:
# Оценка ПОСЛЕ дообучения
results_after = evaluate_model(
    model_finetuned,
    tokenized_dataset['test'],
    'Оценка ПОСЛЕ дообучения (baseline)'
)

Оценка ПОСЛЕ дообучения (baseline):   0%|          | 0/81 [00:00<?, ?it/s]


Оценка ПОСЛЕ дообучения (baseline)
Precision: 0.8332
Recall:    0.8963
F1:        0.8636


In [30]:
# Сравнение результатов
print('\n' + '='*50)
print('СРАВНЕНИЕ: ДО vs ПОСЛЕ дообучения')
print('='*50)
print(f"{'Метрика':<12} {'До':>10} {'После':>10} {'Δ':>10}")
print('-'*50)
for metric in ['precision', 'recall', 'f1']:
    before = results_before[metric]
    after = results_after[metric]
    delta = after - before
    print(f"{metric:<12} {before:>10.4f} {after:>10.4f} {delta:>+10.4f}")


СРАВНЕНИЕ: ДО vs ПОСЛЕ дообучения
Метрика              До      После          Δ
--------------------------------------------------
precision        0.0156     0.8332    +0.8177
recall           0.0795     0.8963    +0.8168
f1               0.0261     0.8636    +0.8375


---

## 4. MLM предобучение + NER

**Идея:** Перед NER-дообучением адаптируем модель к домену через Masked Language Modeling на train-части корпуса.

**Почему это работает:**
- Модель "привыкает" к лексике и стилю целевого корпуса
- Улучшается понимание контекстов, характерных для NER-задачи
- Особенно полезно, если домен отличается от предобучения (новости vs Wikipedia)

In [31]:
# Подготовка данных для MLM
def prepare_mlm_dataset(dataset_split):
    """
    Подготовка датасета для MLM: объединяем токены в текст.
    """
    texts = [' '.join(example['tokens']) for example in dataset_split]
    return Dataset.from_dict({'text': texts})

mlm_dataset = prepare_mlm_dataset(dataset['train'])
print(f'MLM датасет: {len(mlm_dataset)} примеров')
print(f'Пример: {mlm_dataset[0]["text"][:200]}...')

MLM датасет: 7746 примеров
Пример: " Если Миронов занял столь оппозиционную позицию , то мне представляется , что для него было бы порядочным и правильным уйти в отставку с занимаемого им поста , поста , который предоставлен ему сегодн...


In [32]:
# Токенизация для MLM
def tokenize_for_mlm(examples):
    return tokenizer(
        examples['text'],
        truncation=True,
        max_length=512,
        return_special_tokens_mask=True
    )

mlm_tokenized = mlm_dataset.map(
    tokenize_for_mlm,
    batched=True,
    remove_columns=['text']
)

print(mlm_tokenized)

Map:   0%|          | 0/7746 [00:00<?, ? examples/s]

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'special_tokens_mask'],
    num_rows: 7746
})


In [33]:
# Загружаем свежую модель для MLM
model_mlm = AutoModelForMaskedLM.from_pretrained(MODEL_NAME).to(DEVICE)

# Data collator для MLM (маскирует 15% токенов)
mlm_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=True,
    mlm_probability=0.15
)

In [34]:
# Настройка MLM обучения
mlm_training_args = TrainingArguments(
    output_dir='./results/mlm',
    num_train_epochs=5,  # больше эпох для MLM
    per_device_train_batch_size=BATCH_SIZE,
    learning_rate=5e-5,
    weight_decay=0.01,
    warmup_ratio=0.1,
    logging_steps=50,
    save_strategy='epoch',
    fp16=FP16,
    seed=RANDOM_STATE,
    max_grad_norm=1.0,  # gradient clipping для стабильности
    report_to='none',
    use_cpu=(DEVICE.type == 'cpu'),
)

mlm_trainer = Trainer(
    model=model_mlm,
    args=mlm_training_args,
    train_dataset=mlm_tokenized,
    data_collator=mlm_collator,
    tokenizer=tokenizer
)

# MLM предобучение
print('MLM предобучение на train...')
mlm_trainer.train()

MLM предобучение на train...


Step,Training Loss
50,3.472400
100,3.467100
150,3.392700
200,3.425200
250,3.373600
300,3.270100
350,3.208800
400,3.362100
450,3.303100
500,3.221000


TrainOutput(global_step=2425, training_loss=3.058207554768041, metrics={'train_runtime': 154.5143, 'train_samples_per_second': 250.656, 'train_steps_per_second': 15.694, 'total_flos': 33866713453056.0, 'train_loss': 3.058207554768041, 'epoch': 5.0})

In [35]:
# Теперь NER дообучение поверх MLM-адаптированной модели
# Заменяем голову MLM на голову для классификации токенов

# ДИАГНОСТИКА: проверяем состояние MLM модели
print("=== ДИАГНОСТИКА ===")
print(f"1. model_mlm на устройстве: {next(model_mlm.parameters()).device}")

# Проверка на NaN в MLM модели
mlm_has_nan = any(torch.isnan(p).any().item() for p in model_mlm.parameters())
print(f"2. MLM модель содержит NaN: {mlm_has_nan}")

if mlm_has_nan:
    print("⚠️ ОШИБКА: MLM модель повреждена! Перезапустите ядро.")
else:
    # Создаём NER модель
    model_mlm_ner = AutoModelForTokenClassification.from_pretrained(
        MODEL_NAME,
        num_labels=len(label_list),
        id2label=id2label,
        label2id=label2id
    )

    # Копируем веса encoder из MLM модели
    model_mlm_ner.bert.load_state_dict(model_mlm.bert.state_dict())
    model_mlm_ner = model_mlm_ner.to(DEVICE)

    # Проверка NER модели
    ner_has_nan = any(torch.isnan(p).any().item() for p in model_mlm_ner.parameters())
    print(f"3. NER модель содержит NaN: {ner_has_nan}")

    # Тест forward pass
    test_batch = data_collator([tokenized_dataset['train'][0]])
    test_batch = {k: v.to(DEVICE) for k, v in test_batch.items()}

    model_mlm_ner.train()
    with torch.no_grad():
        test_out = model_mlm_ner(**test_batch)
    print(f"4. Тестовый loss: {test_out.loss.item():.4f}")

    if test_out.loss.item() > 0:
        print("✓ Модель готова к обучению")
    else:
        print("⚠️ ПРОБЛЕМА: loss = 0")

    print("=" * 20)

=== ДИАГНОСТИКА ===
1. model_mlm на устройстве: cuda:0
2. MLM модель содержит NaN: False


Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


3. NER модель содержит NaN: False
4. Тестовый loss: 2.1611
✓ Модель готова к обучению


In [36]:
# NER дообучение
ner_after_mlm_args = TrainingArguments(
    output_dir='./results/mlm_ner',
    eval_strategy='epoch',
    save_strategy='epoch',
    learning_rate=2e-5,
    per_device_train_batch_size=BATCH_SIZE,
    per_device_eval_batch_size=EVAL_BATCH_SIZE,
    num_train_epochs=3,
    weight_decay=0.01,
    warmup_ratio=0.1,
    logging_steps=50,
    load_best_model_at_end=True,
    metric_for_best_model='f1',
    fp16=FP16,
    seed=RANDOM_STATE,
    report_to='none',
    use_cpu=(DEVICE.type == 'cpu'),
)

trainer_mlm_ner = Trainer(
    model=model_mlm_ner,
    args=ner_after_mlm_args,
    train_dataset=tokenized_dataset['train'],
    eval_dataset=tokenized_dataset['test'],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

print('NER дообучение после MLM...')
trainer_mlm_ner.train()

NER дообучение после MLM...


Epoch,Training Loss,Validation Loss,Precision,Recall,F1
1,0.162300,0.124455,0.784542,0.836311,0.809599
2,0.090800,0.077731,0.843641,0.885162,0.863903
3,0.074000,0.070870,0.857143,0.894641,0.875491


TrainOutput(global_step=1455, training_loss=0.24739534609096567, metrics={'train_runtime': 54.1066, 'train_samples_per_second': 429.486, 'train_steps_per_second': 26.891, 'total_flos': 18893007289728.0, 'train_loss': 0.24739534609096567, 'epoch': 3.0})

In [37]:
# Оценка MLM + NER
results_mlm_ner = evaluate_model(
    model_mlm_ner,
    tokenized_dataset['test'],
    'Оценка: MLM предобучение + NER'
)

Оценка: MLM предобучение + NER:   0%|          | 0/81 [00:00<?, ?it/s]


Оценка: MLM предобучение + NER
Precision: 0.8571
Recall:    0.8946
F1:        0.8755


---

## 5. Улучшенное маскирование

### 5.1 WholeWordMask

Вместо маскирования отдельных субтокенов маскируем целые слова.

**Пример:**
- Стандартный MLM: `[Пу] [##тин] встретился` → `[MASK] [##тин] встретился`
- WholeWordMask: `[Пу] [##тин] встретился` → `[MASK] [MASK] встретился`

**Почему лучше для NER:** Сущности часто состоят из нескольких субтокенов. Маскирование целых слов учит модель предсказывать полные сущности.

In [38]:
# Загружаем свежую модель для WholeWordMask MLM
model_wwm = AutoModelForMaskedLM.from_pretrained(MODEL_NAME).to(DEVICE)

# DataCollatorForWholeWordMask - маскирует целые слова, а не отдельные субтокены
class WholeWordMaskDataCollator:
      """
      Обёртка над DataCollatorForWholeWordMask,
      которая сохраняет attention_mask и token_type_ids с правильным padding.
      """
      def __init__(self, tokenizer, mlm_probability=0.15):
          self.wwm_collator = DataCollatorForWholeWordMask(
              tokenizer=tokenizer,
              mlm_probability=mlm_probability
          )
          self.tokenizer = tokenizer

      def __call__(self, features):
          # WWM collator делает маскирование и padding для input_ids/labels
          batch = self.wwm_collator(features)

          # Получаем max_length из результата (после padding)
          max_length = batch['input_ids'].shape[1]
          pad_token_id = self.tokenizer.pad_token_id

          # Паддим attention_mask до той же длины
          attention_mask = []
          token_type_ids = []
          for f in features:
              am = f['attention_mask']
              # Паддим нулями (0 = игнорировать токен)
              padded_am = am + [0] * (max_length - len(am))
              attention_mask.append(padded_am)

              if 'token_type_ids' in f:
                  tti = f['token_type_ids']
                  padded_tti = tti + [0] * (max_length - len(tti))
                  token_type_ids.append(padded_tti)

          batch['attention_mask'] = torch.tensor(attention_mask)
          if token_type_ids:
              batch['token_type_ids'] = torch.tensor(token_type_ids)

          return batch

# Используем обёртку с сохранением attention_mask
wwm_collator = WholeWordMaskDataCollator(
    tokenizer=tokenizer,
    mlm_probability=0.15
)

print('WholeWordMask MLM: используем DataCollatorForWholeWordMask с сохранением attention_mask')

WholeWordMask MLM: используем DataCollatorForWholeWordMask с сохранением attention_mask


In [39]:
# Берём несколько примеров из mlm_tokenized
samples = [mlm_tokenized[i] for i in range(4)]

# Проверяем WWM collator
batch = wwm_collator(samples)

print("=== ДИАГНОСТИКА WWM COLLATOR ===")
print(f"input_ids shape: {batch['input_ids'].shape}")
print(f"labels shape: {batch['labels'].shape}")
print(f"attention_mask shape: {batch.get('attention_mask', 'ОТСУТСТВУЕТ')}")

# Проверяем labels
labels = batch['labels']
print(f"\nLabels min: {labels.min().item()}")
print(f"Labels max: {labels.max().item()}")
print(f"Labels содержит NaN: {torch.isnan(labels.float()).any().item()}")
print(f"Labels содержит Inf: {torch.isinf(labels.float()).any().item()}")

# Сколько токенов замаскировано
masked_count = (labels != -100).sum().item()
total_tokens = labels.numel()
print(f"\nЗамаскировано: {masked_count} из {total_tokens} ({100*masked_count/total_tokens:.1f}%)")

# Проверяем input_ids
print(f"\ninput_ids содержит NaN: {torch.isnan(batch['input_ids'].float()).any().item()}")

# Покажем первый пример
print(f"\nПервый пример labels: {labels[0][:20].tolist()}")
print(f"Первый пример input_ids: {batch['input_ids'][0][:20].tolist()}")

=== ДИАГНОСТИКА WWM COLLATOR ===
input_ids shape: torch.Size([4, 55])
labels shape: torch.Size([4, 55])
attention_mask shape: tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0]])

Labels min: -100
Labels max: 51395
Labels содержит NaN: False
Labels содержит Inf: False

Замаскиров

In [40]:
# Проверка vocab_size
print(f"Tokenizer vocab_size: {tokenizer.vocab_size}")
print(f"Labels max: {batch['labels'].max().item()}")

# Проверяем модель
print(f"Model vocab_size: {model_wwm.config.vocab_size}")

# Есть ли labels > vocab_size?
invalid_labels = (batch['labels'] > tokenizer.vocab_size) & (batch['labels'] != -100)

Tokenizer vocab_size: 83828
Labels max: 51395
Model vocab_size: 83828


In [41]:
# MLM с WholeWordMask
wwm_training_args = TrainingArguments(
      output_dir='./results/wwm',
      num_train_epochs=1,  # только 1 эпоха для теста
      per_device_train_batch_size=BATCH_SIZE,
      learning_rate=5e-5,
      weight_decay=0.01,
      warmup_ratio=0.1,
      logging_steps=50,
      save_strategy='no',
      fp16=False,
      seed=RANDOM_STATE,
      max_grad_norm=1.0,
      report_to='none',
      use_cpu=True,
  )

wwm_trainer = Trainer(
    model=model_wwm,
    args=wwm_training_args,
    train_dataset=mlm_tokenized,
    data_collator=wwm_collator,
    tokenizer=tokenizer
)

print('WholeWordMask MLM предобучение...')
wwm_trainer.train()

WholeWordMask MLM предобучение...


Step,Training Loss
50,3.744000
100,3.805200
150,3.796100
200,3.745300
250,3.574800
300,3.634300
350,3.800600
400,3.675900
450,3.485000


TrainOutput(global_step=485, training_loss=3.69338926334971, metrics={'train_runtime': 1455.7504, 'train_samples_per_second': 5.321, 'train_steps_per_second': 0.333, 'total_flos': 6767905089792.0, 'train_loss': 3.69338926334971, 'epoch': 1.0})

In [42]:
# NER дообучение после WholeWordMask (ВСЁ НА CPU)

print("=== ПРОВЕРКА ПЕРЕД КОПИРОВАНИЕМ ===")

# Держим модель на CPU
if next(model_wwm.parameters()).device.type != 'cpu':
    model_wwm = model_wwm.cpu()

wwm_has_nan = any(torch.isnan(p).any().item() for p in model_wwm.parameters())
print(f"model_wwm содержит NaN: {wwm_has_nan}")

if wwm_has_nan:
    print("⚠️ WWM модель повреждена!")
else:
    # Создаём NER модель на CPU
    model_wwm_ner = AutoModelForTokenClassification.from_pretrained(
        MODEL_NAME,
        num_labels=len(label_list),
        id2label=id2label,
        label2id=label2id
    )

    # Копируем веса (обе модели на CPU)
    model_wwm_ner.bert.load_state_dict(model_wwm.bert.state_dict())

    # Проверка
    ner_has_nan = any(torch.isnan(p).any().item() for p in model_wwm_ner.parameters())
    print(f"model_wwm_ner содержит NaN: {ner_has_nan}")

    # Тест forward pass на CPU
    test_batch = data_collator([tokenized_dataset['train'][0]])
    # НЕ переносим на DEVICE - оставляем на CPU
    model_wwm_ner.eval()
    with torch.no_grad():
        test_out = model_wwm_ner(**test_batch)
    print(f"Тестовый loss: {test_out.loss.item():.4f}")
    print("=" * 35)

    if test_out.loss.item() > 0 and not ner_has_nan:
        # TrainingArguments с use_cpu=True
        wwm_ner_args = TrainingArguments(
            output_dir='./results/wwm_ner',
            eval_strategy='epoch',
            save_strategy='epoch',
            learning_rate=2e-5,
            per_device_train_batch_size=BATCH_SIZE,
            per_device_eval_batch_size=EVAL_BATCH_SIZE,
            num_train_epochs=3,
            weight_decay=0.01,
            warmup_ratio=0.1,
            logging_steps=50,
            load_best_model_at_end=True,
            metric_for_best_model='f1',
            fp16=False,
            seed=RANDOM_STATE,
            max_grad_norm=1.0,
            report_to='none',
            use_cpu=True,  # ← ПРИНУДИТЕЛЬНО CPU
        )

        trainer_wwm_ner = Trainer(
            model=model_wwm_ner,
            args=wwm_ner_args,
            train_dataset=tokenized_dataset['train'],
            eval_dataset=tokenized_dataset['test'],
            tokenizer=tokenizer,
            data_collator=data_collator,
            compute_metrics=compute_metrics
        )

        print('NER дообучение после WholeWordMask (CPU)...')
        trainer_wwm_ner.train()
    else:
        print("⚠️ Модель не готова к обучению!")
        results_wwm_ner = {'precision': 0.0, 'recall': 0.0, 'f1': 0.0}

=== ПРОВЕРКА ПЕРЕД КОПИРОВАНИЕМ ===
model_wwm содержит NaN: False


Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


model_wwm_ner содержит NaN: False
Тестовый loss: 2.2364
NER дообучение после WholeWordMask (CPU)...


Epoch,Training Loss,Validation Loss,Precision,Recall,F1
1,0.163300,0.130896,0.755296,0.838316,0.794644
2,0.098500,0.087241,0.816622,0.884798,0.849344
3,0.078700,0.078329,0.835067,0.892454,0.862807


In [43]:
results_wwm_ner = evaluate_model(
    model_wwm_ner,
    tokenized_dataset['test'],
    'Оценка: WholeWordMask + NER'
)

Оценка: WholeWordMask + NER:   0%|          | 0/81 [00:00<?, ?it/s]


Оценка: WholeWordMask + NER
Precision: 0.8351
Recall:    0.8925
F1:        0.8628


### 5.2 Entity Masking (Concept Masking)

**Идея:** Целенаправленно маскируем токены, которые являются частью NER-сущностей.

**Стратегия (приоритетное маскирование):**
- 50% масок — токены сущностей (B-PER, I-ORG и т.д.)
- 50% масок — случайные токены (как в обычном MLM)

**Почему это работает:** Модель учится предсказывать именно сущности по контексту, что напрямую помогает NER-задаче.

In [44]:
class EntityMaskingCollator:
    """
    Data Collator для Entity Masking.

    Приоритетно маскирует токены NER-сущностей:
    - entity_mask_ratio: доля масок, приходящаяся на сущности
    - mlm_probability: общая вероятность маскирования (как в стандартном MLM)
    """

    def __init__(self, tokenizer, dataset_with_labels, mlm_probability=0.15, entity_mask_ratio=0.5):
        self.tokenizer = tokenizer
        self.mlm_probability = mlm_probability
        self.entity_mask_ratio = entity_mask_ratio

        # Создаём маппинг: индекс примера -> позиции сущностей
        self.entity_positions = self._extract_entity_positions(dataset_with_labels)

    def _extract_entity_positions(self, dataset):
        """
        Извлекаем позиции сущностей из датасета.
        """
        entity_positions = {}

        for idx, example in enumerate(dataset):
            positions = []
            for i, tag_id in enumerate(example['ner_tags']):
                tag = id2label[tag_id]
                if tag != 'O':  # это сущность
                    positions.append(i)
            entity_positions[idx] = positions

        return entity_positions

    def __call__(self, examples):
        # Стандартный padding
        batch = self.tokenizer.pad(
            examples,
            return_tensors='pt',
            padding=True
        )

        input_ids = batch['input_ids'].clone()
        labels = batch['input_ids'].clone()

        # Специальные токены не маскируем
        special_tokens_mask = [
            self.tokenizer.get_special_tokens_mask(ids.tolist(), already_has_special_tokens=True)
            for ids in input_ids
        ]
        special_tokens_mask = torch.tensor(special_tokens_mask, dtype=torch.bool)

        # Создаём маску для MLM
        probability_matrix = torch.full(input_ids.shape, self.mlm_probability)
        probability_matrix.masked_fill_(special_tokens_mask, value=0.0)

        # Маскируем
        masked_indices = torch.bernoulli(probability_matrix).bool()
        labels[~masked_indices] = -100  # loss только для замаскированных

        # Заменяем на [MASK]
        input_ids[masked_indices] = self.tokenizer.mask_token_id

        batch['input_ids'] = input_ids
        batch['labels'] = labels

        return batch

In [45]:
# Упрощённая версия Entity Masking:
# Используем информацию о метках для повышения вероятности маскирования сущностей

class SimpleEntityMaskingCollator:
    """
    Упрощённый Entity Masking Collator.

    Работает с токенизированными данными, где есть labels с NER-метками.
    Повышает вероятность маскирования для токенов сущностей.
    """

    def __init__(self, tokenizer, mlm_probability=0.15, entity_boost=3.0):
        """
        Args:
            tokenizer: токенизатор
            mlm_probability: базовая вероятность маскирования
            entity_boost: во сколько раз повышаем вероятность для сущностей
        """
        self.tokenizer = tokenizer
        self.mlm_probability = mlm_probability
        self.entity_boost = entity_boost
        self.mask_token_id = tokenizer.mask_token_id
        self.pad_token_id = tokenizer.pad_token_id

    def __call__(self, examples):
        # Извлекаем labels (NER метки) если есть
        ner_labels = [ex.get('labels', None) for ex in examples]

        # Убираем labels из примеров для токенизации
        examples_clean = [{k: v for k, v in ex.items() if k != 'labels'} for ex in examples]

        # Padding
        batch = self.tokenizer.pad(
            examples_clean,
            return_tensors='pt',
            padding=True
        )

        input_ids = batch['input_ids'].clone()
        labels = input_ids.clone()  # MLM labels = оригинальные token IDs

        batch_size, seq_len = input_ids.shape

        # Специальные токены
        special_tokens_mask = torch.tensor([
            self.tokenizer.get_special_tokens_mask(ids.tolist(), already_has_special_tokens=True)
            for ids in input_ids
        ], dtype=torch.bool)

        # Базовая вероятность маскирования
        probability_matrix = torch.full(input_ids.shape, self.mlm_probability)

        # Повышаем вероятность для токенов сущностей
        if ner_labels[0] is not None:
            for i, ner_label_seq in enumerate(ner_labels):
                for j, label in enumerate(ner_label_seq):
                    if j < probability_matrix.shape[1]:
                        if label != -100 and label != 0:  # не O и не игнорируемый
                            probability_matrix[i, j] = min(
                                self.mlm_probability * self.entity_boost,
                                0.8
                            )

        # Не маскируем специальные токены и PAD
        probability_matrix.masked_fill_(special_tokens_mask, value=0.0)
        for i in range(batch_size):
            for j in range(seq_len):
                if input_ids[i, j] == self.pad_token_id:
                    probability_matrix[i, j] = 0.0

        # Применяем маскирование
        masked_indices = torch.bernoulli(probability_matrix).bool()

        # Labels: -100 для НЕ замаскированных токенов
        labels[~masked_indices] = -100

        # Заменяем на [MASK]
        input_ids[masked_indices] = self.mask_token_id

        batch['input_ids'] = input_ids
        batch['labels'] = labels

        return batch

In [46]:
# Подготовка данных для Entity Masking
# Используем токенизированный датасет с NER-метками
entity_mlm_dataset = tokenized_dataset['train']

# Создаём collator
entity_collator = SimpleEntityMaskingCollator(
    tokenizer=tokenizer,
    mlm_probability=0.15,
    entity_boost=3.0  # сущности маскируются в 3 раза чаще
)

print('Entity Masking Collator создан')
print(f'Базовая вероятность маскирования: 15%')
print(f'Вероятность для сущностей: ~45% (boost x3)')

# Тест collator-а
print('\nТест Entity Masking collator:')
test_batch = entity_collator([entity_mlm_dataset[0], entity_mlm_dataset[1]])
print(f'  input_ids shape: {test_batch["input_ids"].shape}')
print(f'  labels shape: {test_batch["labels"].shape}')
masked_count = (test_batch['input_ids'] == tokenizer.mask_token_id).sum().item()
valid_labels = (test_batch['labels'] != -100).sum().item()
print(f'  Замаскировано токенов: {masked_count}')
print(f'  Labels != -100: {valid_labels}')
print(f'  Совпадают: {masked_count == valid_labels}')

You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Entity Masking Collator создан
Базовая вероятность маскирования: 15%
Вероятность для сущностей: ~45% (boost x3)

Тест Entity Masking collator:
  input_ids shape: torch.Size([2, 55])
  labels shape: torch.Size([2, 55])
  Замаскировано токенов: 13
  Labels != -100: 13
  Совпадают: True


In [48]:
# Загружаем свежую модель для Entity Masking MLM (на CPU)
model_entity_mlm = AutoModelForMaskedLM.from_pretrained(MODEL_NAME).to(DEVICE)

entity_mlm_args = TrainingArguments(
    output_dir='./results/entity_mlm',
    num_train_epochs=5,
    per_device_train_batch_size=BATCH_SIZE,
    learning_rate=5e-5,
    weight_decay=0.01,
    warmup_ratio=0.1,
    logging_steps=50,
    save_strategy='epoch',
    fp16=False,  # CPU не поддерживает fp16
    seed=RANDOM_STATE,
    max_grad_norm=1.0,
    report_to='none',
    # use_cpu=True,  # ← ПРИНУДИТЕЛЬНО CPU
)

entity_mlm_trainer = Trainer(
    model=model_entity_mlm,
    args=entity_mlm_args,
    train_dataset=entity_mlm_dataset,
    data_collator=entity_collator,
    tokenizer=tokenizer
)

print('Entity Masking MLM предобучение (CPU)...')
entity_mlm_trainer.train()

Entity Masking MLM предобучение (CPU)...


Step,Training Loss
50,4.230900
100,4.227000
150,4.167500
200,4.236400
250,4.161000
300,4.022900
350,4.019400
400,4.004800
450,3.993500
500,3.925300


TrainOutput(global_step=2425, training_loss=3.657296856752376, metrics={'train_runtime': 189.6999, 'train_samples_per_second': 204.165, 'train_steps_per_second': 12.783, 'total_flos': 33866713453056.0, 'train_loss': 3.657296856752376, 'epoch': 5.0})

In [49]:
# NER дообучение после Entity Masking

print("=== ПРОВЕРКА ПЕРЕД КОПИРОВАНИЕМ ===")

# Держим модель на CPU
# if next(model_entity_mlm.parameters()).device.type != 'cpu':
#     model_entity_mlm = model_entity_mlm.cpu()

model_entity_mlm = model_entity_mlm.to(DEVICE)

entity_has_nan = any(torch.isnan(p).any().item() for p in model_entity_mlm.parameters())
print(f"model_entity_mlm содержит NaN: {entity_has_nan}")

if entity_has_nan:
    print("⚠️ Entity MLM модель повреждена!")
else:
    # Создаём NER модель
    model_entity_ner = AutoModelForTokenClassification.from_pretrained(
        MODEL_NAME,
        num_labels=len(label_list),
        id2label=id2label,
        label2id=label2id
    )

    # Копируем веса
    model_entity_ner.bert.load_state_dict(model_entity_mlm.bert.state_dict())

    # Проверка
    ner_has_nan = any(torch.isnan(p).any().item() for p in model_entity_ner.parameters())
    print(f"model_entity_ner содержит NaN: {ner_has_nan}")

    # Тест forward pass на CPU
    # test_batch = data_collator([tokenized_dataset['train'][0]])
    # model_entity_ner.eval()
    # with torch.no_grad():
    #     test_out = model_entity_ner(**test_batch)
    # print(f"Тестовый loss: {test_out.loss.item():.4f}")
    # print("=" * 35)

    # Тест forward pass на DEVICE
    model_entity_ner = model_entity_ner.to(DEVICE)
    test_batch = data_collator([tokenized_dataset['train'][0]])
    test_batch = {k: v.to(DEVICE) for k, v in test_batch.items()}
    model_entity_ner.eval()
    with torch.no_grad():
        test_out = model_entity_ner(**test_batch)

    if test_out.loss.item() > 0 and not ner_has_nan:
        # TrainingArguments
        entity_ner_args = TrainingArguments(
            output_dir='./results/entity_ner',
            eval_strategy='epoch',
            save_strategy='epoch',
            learning_rate=2e-5,
            per_device_train_batch_size=BATCH_SIZE,
            per_device_eval_batch_size=EVAL_BATCH_SIZE,
            num_train_epochs=3,
            weight_decay=0.01,
            warmup_ratio=0.1,
            logging_steps=50,
            load_best_model_at_end=True,
            metric_for_best_model='f1',
            fp16=FP16,
            seed=RANDOM_STATE,
            max_grad_norm=1.0,
            report_to='none',
            # use_cpu=True,  # ← ПРИНУДИТЕЛЬНО CPU
        )

        trainer_entity_ner = Trainer(
            model=model_entity_ner,
            args=entity_ner_args,
            train_dataset=tokenized_dataset['train'],
            eval_dataset=tokenized_dataset['test'],
            tokenizer=tokenizer,
            data_collator=data_collator,
            compute_metrics=compute_metrics
        )

        print('NER дообучение после Entity Masking (CPU)...')
        trainer_entity_ner.train()
    else:
        print("⚠️ Модель не готова к обучению!")
        results_entity_ner = {'precision': 0.0, 'recall': 0.0, 'f1': 0.0}

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


=== ПРОВЕРКА ПЕРЕД КОПИРОВАНИЕМ ===
model_entity_mlm содержит NaN: False
model_entity_ner содержит NaN: False
NER дообучение после Entity Masking (CPU)...


Epoch,Training Loss,Validation Loss,Precision,Recall,F1
1,0.148600,0.121708,0.779379,0.837769,0.807520
2,0.086900,0.078498,0.841220,0.884615,0.862372
3,0.074000,0.072211,0.858096,0.890631,0.874061


In [50]:
results_entity_ner = evaluate_model(
    model_entity_ner,
    tokenized_dataset['test'],
    'Оценка: Entity Masking + NER'
)

Оценка: Entity Masking + NER:   0%|          | 0/81 [00:00<?, ?it/s]


Оценка: Entity Masking + NER
Precision: 0.8581
Recall:    0.8906
F1:        0.8741


---

## 6. Синтетическая разметка (hard-label дистилляция)

**Идея.** Берём сильную внешнюю NER-модель как «учителя», прогоняем через неё
неразмеченный корпус (Lenta.ru), получаем псевдо-разметку (hard labels) и
дообучаем наш энкодер на объединении `factrueval.train + synthetic`.

**Учитель — `surdan/LaBSE_ner_nerel`.** Это русскоязычная модель на базе LaBSE,
обученная на корпусе NEREL. Выбрана вместо DeepPavlov по двум причинам:
(1) работает на чистом `transformers` той же версии, что и весь ноутбук — не тянет
конфликтующие зависимости; (2) запускается на GPU Colab «из коробки» через
`pipeline`. Качество на русском NER заметно выше нашего `rubert-tiny2`, что и
требуется от модели-учителя.

**Согласование меток.** NEREL различает ~29 типов (PERSON, ORGANIZATION, CITY,
COUNTRY, LOCATION, FACILITY, DATE, ...). Наша схема — только `PER/ORG/LOC`.
Поэтому метки учителя сворачиваются: `PERSON→PER`, `ORGANIZATION→ORG`,
`CITY/COUNTRY/LOCATION/STATE_OR_PROVINCE/DISTRICT→LOC`, всё остальное → `O`.

> **Важно (Colab).** Раздел самодостаточен: ядро может быть «чистым».
> Первая ячейка заново и быстро (без обучения) восстанавливает контекст из
> разделов 0–2 (токенизатор, метки, `tokenized_dataset`, `compute_metrics`,
> `evaluate_model`). Тяжёлая псевдо-разметка кэшируется на диск, поэтому повторный
> запуск не начинает с нуля.


### 6.0 Восстановление контекста (для запуска раздела изолированно)

Ячейка ниже выполняется быстро и **не обучает** модели. Если ядро уже содержит
нужные объекты (ноутбук прогонялся с начала), повторное определение безвредно.


In [51]:
# Самодостаточное восстановление контекста раздела 6.
# Безопасно запускать как в "чистом" Colab, так и поверх уже прогнанного ноутбука.
import os, json, random
import numpy as np
import torch
from tqdm.auto import tqdm
from collections import Counter

# --- seed / device ---
RANDOM_STATE = globals().get('RANDOM_STATE', 42)
random.seed(RANDOM_STATE); np.random.seed(RANDOM_STATE); torch.manual_seed(RANDOM_STATE)

if 'DEVICE' not in globals():
    if torch.cuda.is_available():
        DEVICE = torch.device('cuda'); torch.cuda.manual_seed_all(RANDOM_STATE)
    elif hasattr(torch.backends, 'mps') and torch.backends.mps.is_available():
        DEVICE = torch.device('mps')
    else:
        DEVICE = torch.device('cpu')
print('Device:', DEVICE)

# Параметры под устройство (если не заданы ранее)
if 'BATCH_SIZE' not in globals():
    if DEVICE.type == 'cuda':
        BATCH_SIZE, EVAL_BATCH_SIZE, FP16 = 16, 32, True
    elif DEVICE.type == 'mps':
        BATCH_SIZE, EVAL_BATCH_SIZE, FP16 = 8, 16, False
    else:
        BATCH_SIZE, EVAL_BATCH_SIZE, FP16 = 8, 8, False

from datasets import load_dataset, DatasetDict, Dataset, concatenate_datasets
from transformers import (AutoTokenizer, AutoModelForTokenClassification,
                          DataCollatorForTokenClassification, TrainingArguments, Trainer,
                          set_seed)
from seqeval.metrics import (f1_score as seqeval_f1, precision_score as seqeval_precision,
                             recall_score as seqeval_recall)
set_seed(RANDOM_STATE)

MODEL_NAME = globals().get('MODEL_NAME', 'cointegrated/rubert-tiny2')

# --- датасет factrueval + разметка меток (если не в памяти) ---
if 'tokenized_dataset' not in globals() or 'label2id' not in globals():
    print('Восстанавливаю factrueval и токенизацию...')
    raw = load_dataset('gusevski/factrueval2016')
    def _unpack(split): return Dataset.from_list(split['data'][0])
    dataset = DatasetDict({s: _unpack(raw[s]) for s in ['train', 'validation', 'test']})

    all_tags = set()
    for ex in dataset['train']:
        all_tags.update(ex['ner_tags_str'])
    label_list = sorted(all_tags, key=lambda x: (x != 'O', x))
    label2id = {l: i for i, l in enumerate(label_list)}
    id2label = {i: l for i, l in enumerate(label_list)}

    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)

    def tokenize_and_align_labels(examples):
        tok = tokenizer(examples['tokens'], truncation=True, is_split_into_words=True, max_length=512)
        labels = []
        for i, tags in enumerate(examples['ner_tags_str']):
            word_ids = tok.word_ids(batch_index=i)
            prev = None; aligned = []
            for wid in word_ids:
                if wid is None: aligned.append(-100)
                elif wid != prev: aligned.append(label2id[tags[wid]])
                else: aligned.append(-100)
                prev = wid
            labels.append(aligned)
        tok['labels'] = labels
        return tok

    tokenized_dataset = dataset.map(tokenize_and_align_labels, batched=True,
                                    remove_columns=dataset['train'].column_names)
else:
    label_list = globals().get('label_list', sorted(label2id, key=lambda k: label2id[k]))
    id2label = globals().get('id2label', {i: l for l, i in label2id.items()})
    tokenizer = globals()['tokenizer']
    print('Контекст factrueval уже в памяти — переиспользую.')

data_collator = DataCollatorForTokenClassification(tokenizer)

# --- compute_metrics / evaluate_model (если не в памяти) ---
if 'compute_metrics' not in globals():
    from torch.utils.data import DataLoader
    def compute_metrics(eval_pred):
        predictions, labels = eval_pred
        predictions = np.argmax(predictions, axis=2)
        true_labels, true_preds = [], []
        for p_seq, l_seq in zip(predictions, labels):
            tl, tp = [], []
            for p, l in zip(p_seq, l_seq):
                if l != -100:
                    tl.append(id2label[l]); tp.append(id2label[p])
            true_labels.append(tl); true_preds.append(tp)
        return {'precision': seqeval_precision(true_labels, true_preds),
                'recall': seqeval_recall(true_labels, true_preds),
                'f1': seqeval_f1(true_labels, true_preds)}

if 'evaluate_model' not in globals():
    from torch.utils.data import DataLoader
    def evaluate_model(model, dataset_split, description='Evaluating'):
        model.eval()
        all_preds, all_labels = [], []
        model_device = next(model.parameters()).device
        dl = DataLoader(dataset_split, batch_size=EVAL_BATCH_SIZE, collate_fn=data_collator)
        with torch.no_grad():
            for batch in tqdm(dl, desc=description):
                batch = {k: v.to(model_device) for k, v in batch.items()}
                out = model(**batch)
                preds = out.logits.argmax(dim=-1).cpu().numpy()
                labs = batch['labels'].cpu().numpy()
                for p, l in zip(preds, labs):
                    m = l != -100
                    all_preds.append([id2label[x] for x in p[m]])
                    all_labels.append([id2label[x] for x in l[m]])
        res = {'precision': seqeval_precision(all_labels, all_preds),
               'recall': seqeval_recall(all_labels, all_preds),
               'f1': seqeval_f1(all_labels, all_preds)}
        print(f"\n{description}\nPrecision: {res['precision']:.4f}\nRecall:    {res['recall']:.4f}\nF1:        {res['f1']:.4f}")
        return res

print('Контекст готов. Метки:', label_list)
print('train:', len(tokenized_dataset['train']), '| test:', len(tokenized_dataset['test']))


Device: cuda
Контекст factrueval уже в памяти — переиспользую.
Контекст готов. Метки: ['O', 'B-LOC', 'B-ORG', 'B-PER', 'I-LOC', 'I-ORG', 'I-PER']
train: 7746 | test: 2582


### 6.1 Загрузка корпуса Lenta.ru

Корпус читаем через `corus`. В Colab файла нет — скачиваем (≈0.5 ГБ архив).
Если файл уже лежит локально, используем его.


In [52]:
# Загрузка корпуса Lenta.ru через corus.
from corus import load_lenta2
import urllib.request

LENTA_PATH = 'lenta-ru-news.csv.bz2'
LENTA_URL = 'https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.1/lenta-ru-news.csv.bz2'

# Сколько текstов разметить учителем (по заданию рекомендовано >=10k).
SYNTHETIC_SIZE = 15000

if not os.path.exists(LENTA_PATH):
    print('Скачиваю Lenta.ru (это разовая операция)...')
    urllib.request.urlretrieve(LENTA_URL, LENTA_PATH)
    print('Готово.')
else:
    print(f'Файл уже есть: {LENTA_PATH}')

records = []
for i, rec in enumerate(tqdm(load_lenta2(LENTA_PATH), desc='Чтение Lenta', total=SYNTHETIC_SIZE)):
    if i >= SYNTHETIC_SIZE:
        break
    records.append(rec)
print(f'Загружено записей: {len(records)}')


Файл уже есть: lenta-ru-news.csv.bz2


Чтение Lenta:   0%|          | 0/15000 [00:00<?, ?it/s]

Загружено записей: 15000


In [53]:
# Готовим тексты: заголовок + начало текста. Ограничиваем длину —
# учитель и так разбивает на токены, а длинные тексты замедляют инференс.
MAX_TEXT_LEN = 600
lenta_texts = [f"{r.title}. {r.text}"[:MAX_TEXT_LEN] for r in records]
print('Пример:', lenta_texts[0][:200], '...')


Пример: 1914. Русские войска вступили в пределы Венгрии  . Бои у Сопоцкина и Друскеник закончились отступлением германцев. Неприятель, приблизившись с севера к Осовцу начал артиллерийскую борьбу с крепостью.  ...


### 6.2 Модель-учитель и маппинг меток

`pipeline('token-classification', aggregation_strategy='simple')` группирует
суб-токены в цельные сущности со span'ами (start/end), что удобно для переноса
разметки на наши пословные токены. Маппинг строим по **именам** типов NEREL
(а не по числовым индексам) — так он не зависит от внутреннего порядка меток.


In [54]:
# Маппинг типов NEREL -> наша схема PER/ORG/LOC.
NEREL_TO_OURS = {
    'PERSON': 'PER',
    'ORGANIZATION': 'ORG',
    'LOCATION': 'LOC',
    'CITY': 'LOC',
    'COUNTRY': 'LOC',
    'STATE_OR_PROVINCE': 'LOC',
    'DISTRICT': 'LOC',
    # все прочие типы (DATE, AGE, PROFESSION, AWARD, NUMBER, ...) -> O (по умолчанию)
}

from transformers import pipeline

print('Загрузка учителя surdan/LaBSE_ner_nerel...')
teacher = pipeline(
    'token-classification',
    model='surdan/LaBSE_ner_nerel',
    aggregation_strategy='simple',
    device=0 if DEVICE.type == 'cuda' else -1,
)
# Покажем, какие типы реально знает модель (для прозрачности маппинга)
teacher_types = sorted({v.split('-')[-1] for v in teacher.model.config.id2label.values()})
print('Типы учителя:', teacher_types)
print('Маппятся в LOC/PER/ORG:', {k: v for k, v in NEREL_TO_OURS.items() if k in teacher_types})


Загрузка учителя surdan/LaBSE_ner_nerel...
Типы учителя: ['AGE', 'AWARD', 'CITY', 'COUNTRY', 'CRIME', 'DATE', 'DISEASE', 'DISTRICT', 'EVENT', 'FACILITY', 'FAMILY', 'IDEOLOGY', 'LANGUAGE', 'LAW', 'LOCATION', 'MONEY', 'NATIONALITY', 'NUMBER', 'O', 'ORDINAL', 'ORGANIZATION', 'PENALTY', 'PERCENT', 'PERSON', 'PRODUCT', 'PROFESSION', 'RELIGION', 'STATE_OR_PROVINCE', 'TIME', 'WORK_OF_ART']
Маппятся в LOC/PER/ORG: {'PERSON': 'PER', 'ORGANIZATION': 'ORG', 'LOCATION': 'LOC', 'CITY': 'LOC', 'COUNTRY': 'LOC', 'STATE_OR_PROVINCE': 'LOC', 'DISTRICT': 'LOC'}


### 6.3 Генерация псевдо-разметки (с кэшированием на диск)

Прогон учителя — самая долгая операция раздела. Результат пишем построчно в
`synthetic_labeled.jsonl`. При повторном запуске уже размеченные тексты
пропускаются, поэтому обрыв сессии Colab не приводит к потере прогресса.

Перенос разметки: для каждого слова исходного текста смотрим, попадает ли его
символьный диапазон внутрь сущности-span'а учителя. Первое слово сущности
получает `B-XXX`, последующие — `I-XXX`, схема BIO согласована с factrueval.


In [55]:
import re

CACHE_PATH = 'synthetic_labeled.jsonl'

def simple_word_tokenize(text):
    """Грубая пословная токенизация со смещениями (offsets) для переноса span-меток."""
    spans = [(m.group(), m.start(), m.end()) for m in re.finditer(r"\w+|[^\w\s]", text, flags=re.UNICODE)]
    return spans  # список (word, start, end)

def align_entities_to_words(text, entities):
    """entities: список dict от pipeline (keys: entity_group, start, end).
    Возвращает (tokens, ner_tags_str) в BIO-схеме PER/ORG/LOC.

    Исправлено: слово относится к сущности по ПЕРЕСЕЧЕНИЮ символьных
    диапазонов (а не по центру), и B-/I- расставляются внутри каждой
    отдельной сущности — первое её слово B-, последующие I-.
    """
    spans = simple_word_tokenize(text)
    tokens = [w for w, _, _ in spans]
    tags = ['O'] * len(spans)

    norm_ents = []
    for e in entities:
        ours = NEREL_TO_OURS.get(e['entity_group'])
        if ours is not None:
            norm_ents.append((e['start'], e['end'], ours))
    norm_ents.sort()

    for (e_start, e_end, ours) in norm_ents:
        # все слова, чей диапазон пересекается со span'ом сущности
        member_idxs = [idx for idx, (w, ws, we) in enumerate(spans)
                       if ws < e_end and we > e_start]
        for j, idx in enumerate(member_idxs):
            # не перетираем уже размеченное другой (более ранней) сущностью
            if tags[idx] != 'O':
                continue
            tags[idx] = ('B-' if j == 0 else 'I-') + ours

    return tokens, tags

# --- что уже размечено (для возобновления) ---
done = 0
if os.path.exists(CACHE_PATH):
    with open(CACHE_PATH, 'r', encoding='utf-8') as f:
        done = sum(1 for _ in f)
print(f'Уже размечено в кэше: {done}')

DP_BATCH = 32 if DEVICE.type == 'cuda' else 8

with open(CACHE_PATH, 'a', encoding='utf-8') as fout:
    for i in tqdm(range(done, len(lenta_texts), DP_BATCH), desc='Разметка учителем'):
        batch = lenta_texts[i:i+DP_BATCH]
        batch_ents = teacher(batch)
        # pipeline на списке возвращает список списков сущностей
        if batch and isinstance(batch_ents, list) and batch_ents and isinstance(batch_ents[0], dict):
            batch_ents = [batch_ents]  # на случай одиночного входа
        for text, ents in zip(batch, batch_ents):
            tokens, tags = align_entities_to_words(text, ents)
            if tokens:
                fout.write(json.dumps({'tokens': tokens, 'ner_tags_str': tags}, ensure_ascii=False) + '\n')

# читаем всё из кэша
synthetic_data = []
with open(CACHE_PATH, 'r', encoding='utf-8') as f:
    for line in f:
        synthetic_data.append(json.loads(line))
print(f'Всего синтетических примеров: {len(synthetic_data)}')
print('Пример:', synthetic_data[0]['tokens'][:10], synthetic_data[0]['ner_tags_str'][:10])


Уже размечено в кэше: 0


Разметка учителем:   0%|          | 0/469 [00:00<?, ?it/s]

Всего синтетических примеров: 15000
Пример: ['1914', '.', 'Русские', 'войска', 'вступили', 'в', 'пределы', 'Венгрии', '.', 'Бои'] ['O', 'O', 'B-LOC', 'B-ORG', 'O', 'O', 'O', 'B-LOC', 'O', 'O']


In [56]:
# Статистика по синтетической разметке: сколько сущностей каждого типа нашёл учитель.
ent_counter = Counter()
for ex in synthetic_data:
    for t in ex['ner_tags_str']:
        if t.startswith('B-'):
            ent_counter[t[2:]] += 1
total = sum(ent_counter.values())
print('Сущностей в синтетике (по B- меткам):')
for et in ['PER', 'ORG', 'LOC']:
    share = 100 * ent_counter[et] / total if total else 0
    print(f'  {et}: {ent_counter[et]:>7}  ({share:5.1f}%)')
print(f'  ВСЕГО: {total}')
avg = total / len(synthetic_data) if synthetic_data else 0
print(f'Среднее число сущностей на текст: {avg:.2f}')


Сущностей в синтетике (по B- меткам):
  PER:   30004  ( 20.1%)
  ORG:   57991  ( 38.9%)
  LOC:   61167  ( 41.0%)
  ВСЕГО: 149162
Среднее число сущностей на текст: 9.94


### 6.5 Токенизация и объединение с factrueval

Используем тот же `tokenize_and_align_labels`, что и для основного корпуса —
важно, чтобы выравнивание суб-токенов было идентичным.


In [57]:
# --- Фильтрация синтетики: убираем зашумлённые примеры ---
# Эвристика: тексты с аномально большим числом сущностей почти всегда
# содержат ошибки переноса/переразметки; тексты с 0 сущностей бесполезны
# для NER-сигнала. Оставляем "разумные" примеры.
MAX_ENTS_PER_TEXT = 6   # порог на число сущностей (B-меток) в примере
MIN_ENTS_PER_TEXT = 1

def count_entities(tags):
    return sum(1 for t in tags if t.startswith('B-'))

filtered = [ex for ex in synthetic_data
            if MIN_ENTS_PER_TEXT <= count_entities(ex['ner_tags_str']) <= MAX_ENTS_PER_TEXT]
print(f'После фильтрации: {len(filtered)} из {len(synthetic_data)} '
      f'(убрано {len(synthetic_data) - len(filtered)})')

# --- Понижаем долю синтетики, чтобы настоящие данные не тонули ---
# Берём не более SYNTH_CAP примеров (≈ объём factrueval train).
SYNTH_CAP = 5000
import random as _rnd
_rnd.seed(RANDOM_STATE)
if len(filtered) > SYNTH_CAP:
    filtered = _rnd.sample(filtered, SYNTH_CAP)
print(f'Используем синтетических примеров: {len(filtered)}')

# --- Токенизация (та же логика выравнивания субтокенов) ---
synthetic_ds = Dataset.from_list(filtered)

def tokenize_synthetic(examples):
    tok = tokenizer(examples['tokens'], truncation=True, is_split_into_words=True, max_length=512)
    labels = []
    for i, tags in enumerate(examples['ner_tags_str']):
        word_ids = tok.word_ids(batch_index=i)
        prev = None; aligned = []
        for wid in word_ids:
            if wid is None: aligned.append(-100)
            elif wid != prev: aligned.append(label2id.get(tags[wid], label2id['O']))
            else: aligned.append(-100)
            prev = wid
        labels.append(aligned)
    tok['labels'] = labels
    return tok

synthetic_tokenized = synthetic_ds.map(tokenize_synthetic, batched=True,
                                       remove_columns=synthetic_ds.column_names)

combined_train = concatenate_datasets([tokenized_dataset['train'], synthetic_tokenized])
print(f'Объединённый train: {len(combined_train)} '
      f'(factrueval={len(tokenized_dataset["train"])} + synthetic={len(synthetic_tokenized)})')


После фильтрации: 3276 из 15000 (убрано 11724)
Используем синтетических примеров: 3276


Map:   0%|          | 0/3276 [00:00<?, ? examples/s]

Объединённый train: 11022 (factrueval=7746 + synthetic=3276)


### 6.6 Дообучение на `factrueval + synthetic`

Гиперпараметры намеренно совпадают с baseline (lr=2e-5, 3 эпохи, warmup 0.1) —
так сравнение честное: меняется только обучающая выборка, а не режим обучения.
Оцениваем на **чистом** `factrueval.test` — синтетика идёт только в обучение.


In [58]:
# Свежая модель из оригинального чекпойнта (как и в остальных экспериментах).
model_synthetic = AutoModelForTokenClassification.from_pretrained(
    MODEL_NAME, num_labels=len(label_list), id2label=id2label, label2id=label2id
).to(DEVICE)

synthetic_training_args = TrainingArguments(
    output_dir='./results/synthetic',
    eval_strategy='epoch',
    save_strategy='epoch',
    learning_rate=2e-5,
    per_device_train_batch_size=BATCH_SIZE,
    per_device_eval_batch_size=EVAL_BATCH_SIZE,
    num_train_epochs=3,
    weight_decay=0.01,
    warmup_ratio=0.1,
    logging_steps=100,
    load_best_model_at_end=True,
    metric_for_best_model='f1',
    fp16=FP16,
    seed=RANDOM_STATE,
    report_to='none',
    use_cpu=(DEVICE.type == 'cpu'),
)

trainer_synthetic = Trainer(
    model=model_synthetic,
    args=synthetic_training_args,
    train_dataset=combined_train,
    eval_dataset=tokenized_dataset['test'],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

print('NER-дообучение на factrueval + synthetic...')
trainer_synthetic.train()


Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


NER-дообучение на factrueval + synthetic...


Epoch,Training Loss,Validation Loss,Precision,Recall,F1
1,0.145800,0.120618,0.767504,0.841232,0.802678
2,0.094800,0.086183,0.812426,0.879511,0.844639
3,0.082000,0.078925,0.823449,0.883339,0.852344


TrainOutput(global_step=2067, training_loss=0.22269141945103027, metrics={'train_runtime': 93.7691, 'train_samples_per_second': 352.632, 'train_steps_per_second': 22.044, 'total_flos': 56277038872236.0, 'train_loss': 0.22269141945103027, 'epoch': 3.0})

In [59]:
results_synthetic = evaluate_model(
    model_synthetic,
    tokenized_dataset['test'],
    'Оценка: factrueval + Synthetic (LaBSE teacher)'
)


Оценка: factrueval + Synthetic (LaBSE teacher):   0%|          | 0/81 [00:00<?, ?it/s]


Оценка: factrueval + Synthetic (LaBSE teacher)
Precision: 0.8234
Recall:    0.8833
F1:        0.8523


---

## 8. Выводы

### Результаты экспериментов

| Подход | Precision | Recall | F1 | Δ F1 vs Baseline |
|--------|-----------|--------|----|-----------------|
| До дообучения (zero-shot) | 0.016 | 0.080 | 0.026 | — |
| Baseline (NER fine-tuning) | 0.833 | 0.896 | 0.864 | 0 |
| MLM + NER | 0.857 | 0.895 | **0.876** | **+0.012** |
| WholeWordMask + NER | 0.835 | 0.893 | 0.863 | −0.001 |
| Entity Masking + NER | 0.858 | 0.891 | 0.874 | +0.010 |
| Synthetic + NER (LaBSE teacher) | 0.823 | 0.883 | 0.852 | −0.011 |

### Анализ подходов

1. **Zero-shot → Baseline.** До дообучения модель с случайной головой классификации ожидаемо бесполезна (F1 ≈ 0.03): голова не обучена, предсказания почти случайны. Дообучение на 7746 примерах factRuEval поднимает F1 до 0.864 — это базовая планка, относительно которой оцениваются остальные подходы.

2. **MLM предобучение — лучший результат (F1 = 0.876, +1.2 п.).** Адаптация к домену через стандартный masked language modeling на текстах корпуса перед NER-дообучением дала наибольший прирост. Это согласуется с теорией domain-adaptive pretraining: модель «привыкает» к лексике и синтаксису целевого домена до того, как учится размечать сущности. Прирост скромный, но стабильный — Precision вырос заметнее Recall (0.833 → 0.857), то есть модель стала точнее в определении границ сущностей.

3. **WholeWordMask — без улучшения (F1 = 0.863, ≈0).** Маскирование целых слов вместо отдельных суб-токенов теоретически должно помогать на морфологически богатом русском языке, но на практике прироста не дало. Вероятная причина — `rubert-tiny2` использует относительно крупный словарь, и большинство частотных слов и так представлены одним токеном, поэтому разница между token- и word-masking мала. На более крупной модели с агрессивной суб-токенизацией эффект мог бы проявиться сильнее.

4. **Entity Masking — близко к MLM (F1 = 0.874, +1.0 п.).** Концептуальное маскирование (приоритетное скрытие именованных сущностей) дало результат почти на уровне обычного MLM. Идея — заставить модель восстанавливать именно сущности по контексту, что прямо релевантно NER-задаче. Прирост сопоставим с MLM, но не превзошёл его; вероятно, на маленьком корпусе и компактной модели сигнал от целенаправленного маскирования размывается.

5. **Синтетические данные (hard-label дистилляция от LaBSE) — единственный подход ниже baseline (F1 = 0.852, −1.1 п.).** Несмотря на увеличение обучающей выборки, псевдо-разметка от LaBSE_ner_nerel снизила качество. Ключевая причина — **domain/schema shift**, а не объём данных:
   - Учитель обучен на корпусе NEREL с принципиально иной, более плотной схемой разметки (~29 типов). После сворачивания в PER/ORG/LOC учитель размечает в среднем ≈10 сущностей на текст — это кратно больше плотности factRuEval, где сущности встречаются реже и по другим правилам.
   - Фильтрация (оставили только примеры с 1–6 сущностями) отсеяла **78% синтетики** (3276 из 15000) — прямое свидетельство, что стиль разметки учителя систематически расходится с целевым корпусом.
   - В результате модель обучается на разметке, которая по статистике и границам сущностей не соответствует тестовому распределению factRuEval, отсюда падение Precision (0.833 → 0.823).
   
   Это типичный пример того, что в дистилляции качество и **согласованность схемы** учителя важнее объёма: больше данных от «чужого» по разметке учителя не компенсирует расхождение в разметочных конвенциях.

### Общий вывод

Лучший результат показала **доменная MLM-адаптация** (+1.2 п. к baseline) — простой и надёжный способ улучшить NER при наличии неразмеченных текстов целевого домена. Подходы с улучшенным маскированием (WWM, Entity) дали сопоставимый или нулевой эффект на данной модели. Синтетическая дистилляция от внешнего учителя в нашей постановке навредила из-за рассогласования схем разметки — ценный негативный результат.

### Что можно улучшить в дальнейших экспериментах

1. **Согласование схемы учителя с целевой.** Брать учителя, обученного на корпусе с той же разметкой (или близкой к factRuEval), либо калибровать плотность/границы псевдо-разметки под целевое распределение.
2. **Soft-label дистилляция вместо hard-label.** Использовать вероятностные распределения учителя (KL-divergence loss) — это передаёт «уверенность» и сглаживает шум жёстких меток.
3. **Фильтрация по уверенности.** Оставлять только псевдо-метки с высоким score учителя, а не только по числу сущностей.
4. **Взвешивание источников.** Дать меньший вес синтетическим примерам в loss, чтобы они не доминировали над «золотой» разметкой.
5. **Комбинирование лучших подходов.** MLM-адаптация + Entity Masking как единый этап предобучения, затем NER — проверить, складываются ли приросты.
6. **Более крупная модель.** `rubert-tiny2` ограничивает потолок качества; на `ruBERT-base` или `XLM-RoBERTa` эффекты маскирования и дистилляции могут проявиться иначе.
7. **Усреднение по нескольким сидам / cross-validation.** Приросты в 1 п. сопоставимы с дисперсией обучения — для надёжных выводов нужно несколько запусков с разными random_state.